onnx/shape_inference/implementation.cc

```diff
@@ -687,9 +687,18 @@ std::vector<const TypeProto*> GraphInferencerImpl::doInferencing(
     const std::vector<const TensorProto*>& inputData) {
   SymbolTable* symbolTable = getSymbolTable();
   int numInputs = int(inputTypes.size());
+  if (getIRVersion() >= 4) {
+    if (g_->input_size() != numInputs) {
+      fail_shape_inference("Graph has ", g_->input_size(), " inputs but ", numInputs, " were provided");
+    }
+  } else {
+    // IR < 4 requires all initializers to be optional inputs
+    // So the number of graph input can be larger than the number of node input 
+    if ( numInputs > g_->input_size()) {
```

Comment:
I suggest that if (numInputs < g_input_size()), then we check to ensure that the missing inputs all have an entry in the initializers? That is, if the graph has inputs [x, y, z] while numInputs is 2, then we must check to make sure that z is in the initializer list.

Reply:
Good idea. I have added this case and written a test for it. Thanks!